####################
#
#  Frederick National Laboratory
#  Susan Huse, susan.huse@@nih.gov
#  October 16, 2018
#
####################
#' Insert columns for EnsemblID and GeneCode
#'
#' Insert separate columns for EnsemblID and gene code into a dataframe with one column containing both IDs
#'
#' Reads a column of concatenated EnsemblIDs and gene codes, as generated by the CCBR RNASeq Pipeline
#' and inserts two new columns, one for each identifier.
#' The colname parameter is used to specify the name of an existing column that contains the concatenated EnsemblID and gene code
#' Use "rownames" if these are the rownames, rather than a dataframe column.
#' 
#' If before is set to TRUE, then the two new columns will be inserted as the first columns in the output dataframe.
#' If before is set to FALSE, then the two new columns will be appened at the end, as the last columns in the output dataframe.
#' 
#' Returns a new dataframe with two additional columns, EnsemblID and GeneCode
#'  
#' @param df input dataframe containing a single column of EnsemblID and gene code
#' @param colname the column in hte input dataframe containing the EnsemblID and gene code.  Default = "rowname" to use the rownames, rather than a data column.
#' @param before insert columns in front of data columns (Default=TRUE)
#' 
#' @return dataframe, with the format provided by limma::topTable
#'
#' @author Susan Huse \email{susan.huse@@nih.gov}
#' @keywords RNASeq Pipeliner limma Ensembl gene code
#'
#' @examples
#' myDF <- insert_EnsemblGene(myDF)
#' myDF <- insert_EnsemblGene(myDF, colname="EnsemblGene", after=TRUE)
#' 
#' @export
insert_EnsemblGene <- function(df, colname="rowname", before=TRUE) {
  if(colname == "rowname") {x <- rownames(df)} else {x=df[,colname]}
  colcnt <- ncol(df)
  
  df$EnsemblID <- split_symbol(x, "E")
  df$GeneCode <- split_symbol(x, "G")
  
  if(before) {
    df <- df[, c(colcnt+1, colcnt+2, 1:colcnt)]
  }
  return(df)
}
